021-server-side-rendering/021-server-side-rendering.md

```diff
@@ -0,0 +1,304 @@
+
+**Problem #3: Using session tokens with the SDK**
+
+Now, you've got a session token, you want to make authenticated requests. For example, getting the user's account details. To set a session token on the server-side, we have to use the `X-Fallback-Cookies` header
+
+```js
+import { Client, Account } from "appwrite";
+
+const client = new Client();
+client.setEndpoint("https://cloud.appwrite.io/v1");
+client.setProject("PROJECT_ID");
+
+const serialisedCookies = JSON.stringify({
+  a_session_PROJECT_ID: sessionToken,
+});
+
+client.headers["X-Fallback-Cookies"] = serialisedCookies;
+
+const account = new Account(client);
+const currentUser = account.get();
+```
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+**Problem #1: Solution A - Return session token in response body**
+
+Return the token within the session object, resulting in the follow example code:
+
+```js
+import { Client, Account } from "appwrite";
+
+const client = new Client();
+client.setEndpoint("https://cloud.appwrite.io/v1");
+client.setProject("PROJECT_ID");
+
+const account = new Account(client);
+const session = await account.createAnonymousSession();
+const sessionToken = session.token;
+```
+
+**Problem #1: Solution B - Include headers within the response**
+
+Return the session cookie in the response headers.
+
+```js
+import { parse } from "set-cookie-parser";
+import { Client, Account } from "appwrite";
+
+const client = new Client();
+client.setEndpoint("https://cloud.appwrite.io/v1");
+client.setProject("PROJECT_ID");
+
+const account = new Account(client);
+const session = await account.createAnonymousSession();
+
+const sessionToken = parse(session.headers["set-cookie"]).find((cookie) =>
+  cookie.name.startsWith("a_session")
+).value;
+```
+
+**Problem #2: Solution A - Temporary OAuth2 token exchange for session**
+
+- Modify the oauth2 flow to include a temporary token in the final redirect.
+- Implement a new endpoint, or extend the existing functionality of the magic URL endpoint, to exchange the temporary token for a session token.
+
+![SSR OAuth2 Flow Sequence Diagram](ssr-oauth-flow.png)
+
+1. Browser makes a GET request to the SSR application.
+2. SSR application makes a GET request to the Appwrite API.
+3. Appwrite API redirects the SSR application to the OAuth2 provider.
+4. The SSR application redirects the client to the OAuth2 provider.
+5. User authenticates with the OAuth2 provider.
+6. OAuth2 provider redirects the browser back to the Appwrite API.
+7. Appwrite API sets the session cookie on the Appwrite domain.
+8. Appwrite API redirects the browser back to the client application. **Now, the redirect URL includes a userId & temporary token.** e.g. `myssrapp.com/oauth2/success?userId=loks0n&token=TEMPORARY_TOKEN`
```

Comment:
Currently OAuth works with 1 line of code.
Wit this, it wont. You will need to add extra logic on page load that looks for those params, and then runs method to finish sign-in process.

Similiar to Magic URL - After sending mail, you need 1 extra page that takes URL parals and finishes the logic.

That's kinda annoying :/ Can we find a way to automate this, somehow? Or simplify?

Reply:
Unfortunately, I can't see a way to simplify this flow: we are at the mercy of browser cookie security.

Here's why:
- We need a place to store a session on the client, and send it with browser GET requests.
- Local storage doesn't work because browsers don't send it with requests. We need to use secure cookies.
- Browsers only send cookies that belong to requested domain.
- Domains may only set cookies for their own domain.

The extra page isn't so much of an issue if we have a universal token -> session endpoint, right?

Reply:
I see. I agree it's better to ask developer to write more code than to have it not work on some setups like Safari or phone devices.

Might be worth checking competitors and see if they have setup this complex. if not, maybe they found some trick.

Reply:
Supabase and Firebase are using the same method, unfortunately.